Add admission metrics for webhooks

jpbetz opened the pull request:
Implements the [Admission Webhooks: Prometheus Metrics](https://docs.google.com/document/d/1rDRrC5MNI2An_FeYx8HAsmlQv1WThh9IBb3_shCTbJA/edit#heading=h.5r7dvq4pv2xm) design.

Fixes: https://github.com/kubernetes/kubernetes/issues/55030

ref: https://github.com/kubernetes/features/issues/492

```release-note
Metrics have been added for monitoring admission plugins, including the new dynamic (webhook-based) ones.
```

jpbetz committed:
Add admission metrics

tallclair commented on staging/src/k8s.io/apiserver/pkg/admission/chain.go:
```diff
@@ -16,22 +16,38 @@ limitations under the License.
 
 // NewChainHandler creates a new chain handler from an array of handlers. Used for testing.
-func NewChainHandler(handlers ...Interface) chainAdmissionHandler {
+func NewChainHandler(handlers ...NamedHandler) chainAdmissionHandler {
```
I think you'll need to update a couple unit tests: 
https://github.com/kubernetes/kubernetes/blob/01f205adf51ca34a527f2bf978e5606d1bb3fd59/plugin/pkg/admission/namespace/autoprovision/admission_test.go#L139
https://github.com/kubernetes/kubernetes/blob/01f205adf51ca34a527f2bf978e5606d1bb3fd59/plugin/pkg/admission/persistentvolume/label/admission_test.go#L81

(See https://github.com/kubernetes/kubernetes/pull/55086/files#diff-97778d709e3ef8a841d3749d2c16072a)

jpbetz replied:
Thanks @tallclair! I've fixed the tests, making heavy use of the work you did for #55086.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/metrics.go:
```diff
@@ -0,0 +1,161 @@
+// namedHandler requires each admission.Interface be named, primarly for metrics tracking purposes.
+type NamedHandler interface {
+	Interface
+	GetName() string
```
Is it much harder to just add `Name()` to the existing interface?

jpbetz replied:
Please correct me if I'm wrong, but my current understanding is that since we don't guarantee a 1:1 mapping of `Interface` to plugin, we can't name the interface implementations directly, but must instead supplement them with a name only after they are bound to a plugin, which is what I've gone for here.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/metrics.go:
```diff
@@ -0,0 +1,161 @@
+	latencyBuckets       = prometheus.ExponentialBuckets(10000, 2.0, 8)
+	latencySummaryMaxAge = 5 * time.Hour
+
+	// Admission step metrics. Each step is identified by a distinct type label value.
```
Define 'step'?

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/metrics.go:
```diff
@@ -0,0 +1,161 @@
+	stepMetrics.observe(elapsed, rejected, string(attr.GetOperation()), gvr.Group, gvr.Version, gvr.Resource, attr.GetSubresource(), stepType)
+}
+
+// ObserveAdmissionController records admission related metrics for a build-in admission controller, identified by it's plugin handler name.
```
nit: s/build-in/built-in/

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/metrics.go:
```diff
@@ -0,0 +1,161 @@
+	prometheus.MustRegister(m.latenciesSummary)
+}
+
+func (m *admissionMetrics) observe(elapsed time.Duration, rejected bool, labels ...string) {
```
is the order of the labels necessary?

jpbetz replied:
Yeah, the labels are positional. They're defined using a `[]string` in calls to functions like `NewCounterVec` and are assigned values via the varadic `WithLabelValues` function. It's not type safe, which is why I've hidden this `observe` method behind strongly typed `Observe...()` functions.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/metrics.go:
```diff
@@ -0,0 +1,161 @@
+}
+
+// ObserveExternalWebhook records admission related metrics for a external admission webhook.
+func ObserveExternalWebhook(elapsed time.Duration, rejected bool, hook *v1alpha1.ExternalAdmissionHook, attr Attributes) {
```
What if you make this a method of the `admissionMetrics` object?

Then you can expose the global one by saying

```go
var (
  ObserveExternalWebhook = externalWebhookMetrics.observeExternal
  ...
)
```

Then you can test the internal method and be confident that the exposed one works, and you don't have to reference the global in the test.

jpbetz replied:
I like the idea.

The catch here is that `admissionMetrics` is common to all the different use cases (step, controller, external webhook), but the `Observe*()` functions customized for each of those use cases for type safety purposes:

```
ObserveAdmissionStep(elapsed time.Duration, rejected bool, attr Attributes, stepType string)
ObserveAdmissionController(elapsed time.Duration, rejected bool, handler NamedHandler, attr Attributes)
```
This is why only the `observe()` function, which is common to all the use cases, is on the `admissionMetrics` type.

But I should be able to restructure this to get the test-ability improvement you're suggesting. I'll work on it.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/metrics.go:
```diff
@@ -0,0 +1,161 @@
+	m.latenciesSummary.WithLabelValues(labels...).Observe(elapsedMicroseconds)
+}
+
+func newAdmissionMetrics(name string, labels []string, helpTemplate string) *admissionMetrics {
```
Thank you for making this a constructor which emits a testable thing! I have one suggestion above to make it even more testable.

lavalamp commented:
Can you add a test? Want to make sure that, if anyone ever changes the set of labels, a test at least warns them that monitoring might depend on them.

tallclair commented:
Testing prometheus metrics is a bit messy, but see https://github.com/kubernetes/kubernetes/pull/55086/files#diff-486060c5d6c3f596cc0ea1d054b9a2d6R191 for a starting point.

jpbetz committed:
Add system namespaces to admission metrics. Add tests and leverage test code from PR#55086

jpbetz commented:
/retest pull-kubernetes-bazel-test

jpbetz commented:
/retest pull-kubernetes-bazel-test

jpbetz commented:
/retest pull-kubernetes-bazel-test

jpbetz commented:
/retest pull-kubernetes-bazel-test

jpbetz commented:
/retest pull-kubernetes-bazel-test

jpbetz commented:
/test pull-kubernetes-bazel-test

jpbetz commented:
Rebasing to see if that resolves the issue with `plugin/pkg/admission/namespace/autoprovision/BUILD`, which was not modified by this PR. I've also re-run `hack/update-bazel.sh` (and `hack/update-all.sh`, just to be exhaustive).

jpbetz commented:
Fixed a data race in TestCRD that I repeatedly hit in testing (9903bc3). Not sure why this PR is the first to hit it.

jpbetz committed:
Fix admission metrics to track mutating/validating correctly

Also update admission test mocks to better reflect typical usage and fix broken tests.

jpbetz commented:
@lavalamp @caesarxuchao @tallclair Tests added, PTAL.

Only big remaining question I have is:  Should I add a label for namespace or go with a boolean like isSystemNs?  If I go with the namespace, a kube installation with a large # of namespace might end up with an explosion of actual metrics (metric name + label values), but if I go with isSystemNs, do we get enough granularity?  I'm leaning toward isSystemNs, but feedback is more than welcome!

jpbetz commented:
/retest

tallclair commented on staging/src/k8s.io/apiserver/pkg/admission/chain.go:
```diff
@@ -16,22 +16,45 @@ limitations under the License.
+
 // Admit performs an admission control check using a chain of handlers, and returns immediately on first error
 func (admissionHandler chainAdmissionHandler) Admit(a Attributes) error {
+	var err error
```
nit: this is fragile since it's easy to accidentally mask this with
```
err := foo()
return err
```
But if you make it a named return instead, you don't need to worry about that, e.g. `Admit(a Attributes) (err error) { ...`

Same applies below (Validate)

lavalamp replied:
I don't follow this suggestion. the only difference between starting off with `var err error` and using a named return parameter is that in the former case, empty `return` won't work? Either way it can be shadowed in a nested {} block.

Using a named parameter saves a line of code.

If anything, it is the entire pattern of maintaining state in a variable throughout the entire function to return it at the end that is suspect. I believe the purpose is to defer the Observe call; but that could be handled differently:

```go
finish := func(err error) error {
  Metrics.ObserveAdmissionStep(time.Since(start), err != nil, a, stepMutating)
  return err
}

// And then instead of returning an error directly, always:
return finish(err)

```

tallclair replied:
It gets masked in the scope of the variable, but the defered function still sees the return result: https://gist.github.com/tallclair/b2e6d2217391cbfeafd919bb2f842b46

jpbetz replied:
Good catch @tallclair 

How about splitting this into two functions to avoid maintiaining the state throughout the entire function?

```go
func (admissionHandler chainAdmissionHandler) Validate(a Attributes) error {
     start := time.Now()
     err := admissionHandler.validate(a)
     Metrics.ObserveAdmissionStep(time.Since(start), err != nil, a, stepValidating)
     return err
 }

 func (admissionHandler chainAdmissionHandler) validate(a Attributes) (err error) {
    // original, uninstrumented implementation
  }
```

tallclair replied:
works for me

tallclair commented on staging/src/k8s.io/apiserver/pkg/admission/chain_test.go:
```diff
@@ -17,145 +17,159 @@ limitations under the License.
+	rejectedLabelFilter := map[string]string{"rejected": "true"}
+	for k, v := range labelFilter {
+		rejectedLabelFilter[k] = v
+	}
```
You should add a `"rejected": "false"` label for the accept case, otherwise a rejected count would be included.

jpbetz replied:
Ah, yes. Thanks!

tallclair commented:
> Only big remaining question I have is: Should I add a label for namespace or go with a boolean like isSystemNs?

We're duplicating this conversation... but to copy my answer from the design doc: you don't want label sets to be unbounded, which is why I suggested hardcoding the system vs other split. The only alternative I see is to make it a configurable option in the apiserver (since anything other than kube-system is deployment specific)

lavalamp commented:
Is it important to know if it's the system namespace or not? How about we omit that completely until we're sure we know what we want?

e.g. maybe it'd be better to record the run-level label value, if the namespace has one.

tallclair commented:
> Is it important to know if it's the system namespace or not? How about we omit that completely until we're sure we know what we want?

The goal is to identify a misconfiguration that can bring down the cluster. For a model like GKE, in kube-system vs not seems like the least noisy signal we can get. Users could still try and schedule things in the kube-system namespace, but usually RBAC should prevent that.

> e.g. maybe it'd be better to record the run-level label value, if the namespace has one.

I hadn't heard of run-levels.  Are they already implemented? Is the set of runlevels finite (and small)?

crassirostris commented on staging/src/k8s.io/apiserver/pkg/admission/metrics.go:
```diff
@@ -0,0 +1,165 @@
+func newAdmissionMetrics() *AdmissionMetrics {
+	// Admission metrics for a step of the admission flow. The entire admission flow is broken down into a series of steps
+	// Each step is identified by a distinct type label value.
+	step := newMetricSet("step_",
```
I would recommend to pass the name without underscore and insert it later, to avoid the situation when someone passes a name without an underscore later

lavalamp replied:
Makes sense. I'd also test that the full set of constructed names is what you expect, because if that changes monitoring systems will break.

crassirostris commented on staging/src/k8s.io/apiserver/pkg/admission/metrics.go:
```diff
@@ -0,0 +1,165 @@
+	m.latenciesSummary.WithLabelValues(labels...).Observe(elapsedMicroseconds)
+}
+
+func newMetricSet(name string, labels []string, helpTemplate string) *metricSet {
```
It's easier to parse the code when constructor is close to the type definition. If you don't have a strong preference, please move it there

crassirostris commented on staging/src/k8s.io/apiserver/pkg/admission/metrics.go:
```diff
@@ -0,0 +1,165 @@
+	m.latenciesSummary.Reset()
+}
+
+func (m *metricSet) observe(elapsed time.Duration, labels ...string) {
```
Usually methods are capitalized, since the type is not exported anyway

jpbetz replied:
Will do. Is there a reason for this? I can imagine one advantage might be that if one were to export the type later, that it would require fewer code changes. The down side is that one might then accidentally expose methods that should have been kept unexposed when the type was exposed..

lavalamp replied:
I don't agree. Exporting the method can allow the method to be called if an outside package gets a variable of this type. Don't do it unless you're purposely implementing an interface.

crassirostris replied:
At the same time it creates ambiguity whether this method is meant to be "private" or "public" and whether it's safe it use it outside of other type methods

> Exporting the method can allow the method to be called if an outside package gets a variable of this type

This is rarer situation

jpbetz replied:
The point @lavalamp makes about interfaces is compelling. I'll keep them private.

lavalamp replied:
public/private in go is always package level, not type level (as in C++). If you want stuff protected from other callers in the same package, *you need to split into multiple packages*.

crassirostris commented on staging/src/k8s.io/apiserver/pkg/admission/metrics.go:
```diff
@@ -0,0 +1,165 @@
+}
+
+// namedHandler requires each admission.Interface be named, primarly for metrics tracking purposes.
+type NamedHandler interface {
```
Comment mismatch: `namedHandler` vs `NamedHandler`

Also IMO it's better to define types before methods

lavalamp replied:
You can't have *both* all types before all methods *and* all constructors next to their types (as you request in another comment). ;)

(In this case I agree that NamedHandler ought to go near the top of this file.)

jpbetz replied:
Thanks, I've reorganized the types and updated the docs.

crassirostris replied:
Sorry, to clarify, I meant it's better to define types not in-between other type's methods, not that all types should go before all methods

crassirostris reviewed (commented):
Please add a comment for all exported types/methods/variables

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/metrics.go:
```diff
@@ -0,0 +1,165 @@
+	Metrics = newAdmissionMetrics()
+)
+
+type AdmissionMetrics struct {
```
Godoc please.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/metrics.go:
```diff
@@ -0,0 +1,165 @@
+
+func newMetricSet(name string, labels []string, helpTemplate string) *metricSet {
+	return &metricSet{
+		total: prometheus.NewCounterVec(
```
hm, do you need both a counter and a histogram? Surely a count is constructable from a histogram? Or is it really lossy or something?

jpbetz replied:
Good question. I had based this loosely on the [endpoint metrics](https://github.com/kubernetes/kubernetes/blob/30f015a6fc13979193a6105890ce91ace651eb5c/staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go), which have a counter, histogram and summary.

After reading the docs for each, I don't see the need for a separate counter, you can get that from the summary or histogram.

And even having both histogram and summary is a bit redundant. The main argument to keep both is that the summary has a relatively short  time window and don't aggregate well.  (https://prometheus.io/docs/practices/histograms/) I'll remove this counter.

lavalamp commented:
@tallclair I am confused about why this helps-- what action will the monitoring system perform differently for rejections to kube-system objects?

I think kube-system is not that special and we should not hard code it like this. Run levels are described in #54522. I imagine there will be 5 or fewer.

jpbetz commented:
Updated based on feedback. Only remaining issue is `kube-system` namespace identification issue. @tallclair PTAL.

jpbetz committed:
Remove is_system_ns from admission metrics

jpbetz commented:
@lavalamp, @tallclair, @crassirostris this is ready for another round of review.

I've removed `is_system_ns`. Once we agree on an approach, we can submit a separate PR.

tallclair commented:
Discussed offline. Removing the system namespace label SGTM.

/lgtm

jpbetz commented:
Latest commit was only for `update-bazel.sh`, reapplying existing lgtm.

lavalamp commented:
@jpbetz what do you think about rebasing on top of #55127? I think Chao has more in the critical path than you do.

jpbetz commented:
@lavalamp I'm okay with that.

tallclair commented:
This PR is a cherrypick candidate. Will rebasing on #55127 cause problems with the cherrypick (or is #55127 being cherrypicked too?)

lavalamp commented:
You want to cherrypick this? Why?

lavalamp commented:
I guess @jpbetz could cherrypick and then rebase the branch...

tallclair commented:
> You want to cherrypick this? Why?

Because we don't currently have any monitoring of the admission controllers that are used in production...

jpbetz commented:
This PR is already based atop multiple PRs that were merged after 1.8 was cut, one more won't really make that much difference. But since we're going to monitor pod security via a separate monitoring change, do we still need to backport this? It's basically just new feature development at that point, and unless it fixes a security issue or critical bug, it's difficult to justify cherrypicking feature dev.

caesarxuchao commented on staging/src/k8s.io/apiserver/pkg/admission/metrics.go:
```diff
@@ -0,0 +1,165 @@
+// ObserveAdmissionController records admission related metrics for a built-in admission controller, identified by it's plugin handler name.
+func (m *AdmissionMetrics) ObserveAdmissionController(elapsed time.Duration, rejected bool, handler NamedHandler, attr Attributes, stepType string) {
+	gvr := attr.GetResource()
+	m.controller.observe(elapsed, handler.Name(), stepType, string(attr.GetOperation()), gvr.Group, gvr.Version, gvr.Resource, attr.GetSubresource(), isSystemNsLabel(attr), strconv.FormatBool(rejected))
```
I think you might have already checked, does `gvr.Resource` contains the `subresource` already?

caesarxuchao commented on staging/src/k8s.io/apiserver/pkg/admission/metrics.go:
```diff
@@ -0,0 +1,165 @@
+}
+
+// ObserveExternalWebhook records admission related metrics for a external admission webhook.
+func (m *AdmissionMetrics) ObserveExternalWebhook(elapsed time.Duration, rejected bool, hook *v1alpha1.ExternalAdmissionHook, attr Attributes) {
```
Remove the `External` in the method name.

caesarxuchao commented on staging/src/k8s.io/apiserver/pkg/admission/metrics.go:
```diff
@@ -0,0 +1,151 @@
+}
+
+// ObserveExternalWebhook records admission related metrics for a external admission webhook.
+func (m *AdmissionMetrics) ObserveExternalWebhook(elapsed time.Duration, rejected bool, hook *v1alpha1.Webhook, attr Attributes) {
```
Please remove `External` from the method name.

jpbetz replied:
Simplifying to `webhook` everywhere.

caesarxuchao reviewed (commented):
A few nits. Addressing them in follow-up pulls is fine.

lavalamp commented:
Yeah this seems like kind of a major thing to cherry pick.

caesarxuchao commented on staging/src/k8s.io/apiserver/pkg/admission/chain_test.go:
```diff
@@ -17,145 +17,171 @@ limitations under the License.
+		if len(calls) > 0 {
+			if len(calls) > 1 {
+				// When not accepted, ensure that all but the last controller had been accepted, since
+				// the chain stops execution at the first rejection.
```
Clarified offline. I had misread.

lavalamp commented:
/approve

crassirostris commented on staging/src/k8s.io/apiserver/pkg/admission/metrics.go:
```diff
@@ -0,0 +1,151 @@
+			prometheus.HistogramOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      fmt.Sprintf("%s_latencies", name),
```
Should be `fmt.Sprintf("%s_request_latencies_microseconds")`, see https://prometheus.io/docs/practices/naming/. It's currently super unclear what `admission_step_latencies` mean, maybe I'm even wrong and those are not requests which only proves my point :)

Additionally we have bunch of metrics with different units, so adding it to the metric name is also a good idea

jpbetz replied:
We originally had the  unit on the metric names and subsequently removed them. This was discussed in the design doc (https://docs.google.com/document/d/1rDRrC5MNI2An_FeYx8HAsmlQv1WThh9IBb3_shCTbJA/edit#bookmark=id.z2fplna0fwhs). The decision was to favor consistency with existing kube-apiserver metric names already in use.

Each of metrics are on a "admission" operation.

crassirostris replied:
TBH, I don't see any actual discussion and there are several metrics in this repo which expose the metrics in a correct way already:

https://github.com/kubernetes/kubernetes/blob/master/pkg/client/metrics/prometheus/prometheus.go
https://github.com/kubernetes/kubernetes/blob/master/pkg/util/reflector/prometheus/prometheus.go

I don't think being consistent with some subset of metrics is a good enough reason to ditch the guidelines. FWIW, it's a good way to gradually enforce the standard

jpbetz replied:
I'm happy to apply guidelines if we're aspiring toward a specific set of them. Is the guidance from sig-instrumentation to follow the prometheus naming guidelines? Any supplemental guidelines we following?  My understanding of the prometheus naming guide is that I should have something like: "%_<event>_latencies_seconds", does that sound right?

jpbetz replied:
I've gone with _seconds_ based on my understanding of prometheus naming conventions.

crassirostris replied:
Thanks a lot! We're always happy to help, feel free to ping me or @piosz :)

> Is the guidance from sig-instrumentation to follow the prometheus naming guidelines?

Yes

> Any supplemental guidelines we following? 

Not yet. As with the logging verbosity levels, this is a problem nobody had time to work on :(

> My understanding of the prometheus naming guide is that I should have something like: "%__latencies_seconds", does that sound right?

SGTM, from working with other metrics, that seems like a great idea, to use seconds instead (nano|micro|milli)seconds

crassirostris commented on staging/src/k8s.io/apiserver/pkg/admission/metrics.go:
```diff
@@ -0,0 +1,151 @@
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      fmt.Sprintf("%s_latencies", name),
+				Help:      fmt.Sprintf(helpTemplate, "latency histogram"),
```
Same question as for the metric name, latency of what?

jpbetz replied:
Latency of *admission*. See answer above.

crassirostris replied:
Then please swap the subject and action, i.e. not `admission_controller_latencies`, but `controller_admission_latencies`, because it's not `request_client_latencies` but `client_request_latencies`

crassirostris replied:
Maybe you don't even need it in the name, you can make a label out of it

Anyways, next time you have some doubts, please ask someone from the sig-intrumetation to participate in the discussion

/cc @piosz @fgrzadkowski

jpbetz replied:
Label is a good idea, I've re-purposed the 'type' label slightly so that it is either _validate_ or _admit_, which should clarify this.

Thanks for the review!

crassirostris replied:
Thank you! That really helps!

crassirostris commented:
One last nit with the metric names/help strings

jpbetz committed:
Align admission metric names with prometheus guidelines

jpbetz commented:
I've rebased and applied @crassirostris and @caesarxuchao's feedback.

crassirostris commented on staging/src/k8s.io/apiserver/pkg/admission/metrics.go:
```diff
@@ -0,0 +1,151 @@
+			prometheus.HistogramOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      fmt.Sprintf("%s_latencies_seconds", name),
```
I still think the metric name is a wrong place to make the distinction, essentially this is the same metric just from different places. `Subsystem` is a much more suitable name for this case

To me, `adminssions_controller_admissions_latencies_seconds`  sounds better than `adminssions_controller_latencies_seconds`, because it's clear what's the metric is about. If a new metric appears, e.g. `adminssions_controller_doing_stuff_latencies_seconds`, it won't be consistent. Additionally, when exporting to other systems, e.g. Stackdriver Monitoring, making metric name itself self-contained makes even more sense

crassirostris replied:
@fabxc @piosz WDYT? Let's have some quorum here

crassirostris replied:
Also @fgrzadkowski

We should totally create K8s guidelines for metrics

jpbetz replied:
`adminssions_controller_admissions_latencies_seconds` is what we had originally proposed in the design before adjustments based on feedback, so if we can get agreement on that as an instrumentation naming convention going forward, I'm more than happy to change it back.

jpbetz replied:
@crassirostris I don't understand the name vs. subsystem suggestion. What are you proposing?

crassirostris replied:
Effectively that would be the same exposed name, but in the code that would look like this:

```go
prometheus.NewHistogramVec(
	prometheus.HistogramOpts{
		Namespace: namespace,
		Subsystem: fmt.Sprintf("admission_%s", name),
		Name:      "admission_latencies_seconds",
		Help:      fmt.Sprintf(helpTemplate, "latency histogram"),
		Buckets:   latencyBuckets,
	},
	labels,
),
```

So it looks like the same metric, but for different components: admission webhook, admission controller, admission step.

Another way I find cleaner than the current solution would be to have a metric named `apiserver_admission_latencies_seconds` that has a label `component` or `source`

jpbetz replied:
I've updated the metric names to include _admission_.

We need to keep the existing conmponents separate to adhere to the "[metric names] should represent the same logical thing-being-measured across all label dimensions" prometheus guideline. In this case, steps, controllers and webhooks are all logically distinct-- components are compiled in code while webhooks are called remotely... If combined, a unfiltered query against the combined metric is at risk of being more misleading than helpful, also suggesting they should be kept logically separate.

Subsystem does not strike me as a good field for component names. `admission` is the logical subsystem here, as far as I can tell the hierarchy we use for metrics is something like service/subsystem/component/metric.  Which suggests keeping subsystem as is. It's also an implementation detail and doesn't impact the exposed metric name.

crassirostris replied:
Thanks!

> If combined, a unfiltered query against the combined metric is at risk of being more misleading than helpful

Sounds reasonable

crassirostris reviewed (commented):
Thanks! One last thing, everything else LGTM

crassirostris commented:
Metric names seem more understandable now, at least to me, a person without much context, thanks a lot!

/lgtm

jpbetz commented:
/test pull-kubernetes-kubemark-e2e-gce

deads2k commented on staging/src/k8s.io/apiserver/pkg/admission/chain.go:
```diff
@@ -16,22 +16,46 @@ limitations under the License.
+
+// chainAdmissionHandler is an instance of admission.NamedHandler that performs admission control using
+// a chain of admission handlers
+type chainAdmissionHandler []NamedHandler
```
@jpbetz This type is a union type of `Interfaces` to produce a union `Interface`.  This change breaks the composition model of having a union type that unions the types it is exposing.  Please make a followup that restores the composition model.

You can do this by creating an `Interface` wrapper which provides both `Admit` and `Validate`, and does type casting on the delegate.  That will preserve the abstractions and allows layering.

deads2k commented on plugin/pkg/admission/namespace/autoprovision/admission_test.go:
```diff
@@ -129,19 +129,15 @@ func TestAdmissionNamespaceExists(t *testing.T) {
-	err = chainHandler.Admit(admission.NewAttributesRecord(&pod, nil, api.Kind("Pod").WithVersion("version"), pod.Namespace, pod.Name, api.Resource("pods").WithVersion("version"), "", admission.Update, nil))
-	if err != nil {
-		t.Errorf("unexpected error returned from admission handler")
+	if handler.Handles(admission.Update) {
```
@jpbetz This check isn't the same as the `Admit` call that was previously done.  You need a followup to restore the test.

deads2k commented:
@jpbetz looks like your forgot to tag @kubernetes/sig-api-machinery-pr-reviews and/or myself, @sttts, or @derekwaynecarr (owners of the local admission area).  I've made a couple comments that need to be addressed.  Please make the requested changes and copy us into the followup pulls.

jpbetz commented:
@deads2k, my mistake, I'll submit a follow up PR shortly.

smarterclayton commented:
This is O(M*N) with resources and admission controllers.  On a large cluster with lots of CRDs this causes a significant amount of memory use and churn.  On a smaller cluster with lots of admission controllers and CRDs, 75% of all apiserver prometheus metrics (38k series out of 46k series) were this metric.

The primary dimension that is arguably not useful is by resource type and sub resource, since the vast majority of these are not matched.  

I think we should drop the resource and subresource latency tracking.

tallclair commented:
@smarterclayton There are 3 different metrics added in this PR: "step", "controller", and "webhook". I believe controller & webhook are only recorded when the admission controller handles the resource type. Step is for all resources, but only not by controller. Do you know what the breakdown of these 3 metrics are in the clusters you're looking at? Are they all problematic, or just one?
